src/libertem/analysis/gridmatching.py

```diff
@@ -0,0 +1,303 @@
+    cutoff = parameters['max_delta'] * parameters["tolerance"]
+    indices = vector_solver(points, zero, a, b)
+    rounded = np.around(indices)
+    errors = np.linalg.norm(np.absolute(indices - rounded), axis=1)
+    # FIXME magic number, make user parameter
+    matched_selector = errors < cutoff
+    matched = points[matched_selector]
+    matched_indices = rounded[matched_selector].astype(np.int)
+    remainder = points[np.invert(matched_selector)]
+    return (matched, matched_indices, remainder)
+
+
+# Try out all combinations of polar_vectors and
+# see which one generates most matches
+def do_match(points, zero, polar_vectors, parameters):
+    match_matrix = {}
+    # we test all pairs of candidate vectors
+    # and populate match_matrix
+    for i in range(len(polar_vectors)):
+        for j in range(i + 1, len(polar_vectors)):
+            a = polar_vectors[i]
+            b = polar_vectors[j]
+            # too parallel, not good lattice vectors
+            if not angle_check(np.array([a]), np.array([b]), parameters['min_angle']):
+                continue
+            a, b = make_cartesian(np.array([a, b]))
+
+            match = match_all(points, zero, a, b, parameters)
+            # At least three points matched
+            if len(match[0]) > 2:
+                match_matrix[(i, j)] = match
+    return match_matrix
+
+
+# Find the pair of lattice vectors that generates most matches
+# Return vectors and additional info about the match
+def find_best_vector_match(points, zero, candidates, parameters):
+    match_matrix = do_match(points, zero, candidates, parameters)
+    if match_matrix:
+        # we select the entry with highest number of matches
+        candidate_index, match = sorted(
+            match_matrix.items(), key=lambda d: len(d[1][1]), reverse=True
+        )[0]
+        polar_a = candidates[candidate_index[0]]
+        polar_b = candidates[candidate_index[1]]
+        a, b = make_cartesian(np.array([polar_a, polar_b]))
+        return (a, b, match)
+    else:
+        raise NotFoundException
+
+
+def optimize(zero, a, b, matched, matched_indices):
+    # The minimizer expects a flat array, so we disassemble the vectors
+    z1, z2 = zero
+    a1, a2 = a
+    b1, b2 = b
+
+    def error_function(vec):
+        # ... and reassemble them here
+        z1, z2, a1, a2, b1, b2 = vec
+        zero = np.array((z1, z2))
+        a = np.array((a1, a2))
+        b = np.array((b1, b2))
+        errors = matched - calc_coords(zero, a, b, matched_indices)
+        error = (errors ** 2).sum()
+        return error
+
+    x0 = np.array((z1, z2, a1, a2, b1, b2))
+    res = scipy.optimize.minimize(error_function, x0=x0)
+
+    z1, z2, a1, a2, b1, b2 = res.x
+
+    new_zero = np.array((z1, z2))
+    new_a = np.array((a1, a2))
+    new_b = np.array((b1, b2))
+
+    return np.array((new_zero, new_a, new_b))
+
+
+# In the real world, this would distinguish more between finding good candidates
```

Comment:
These comments would actually be better as doc strings, especially for the `full_match` and `fast_match` functions, which are the entry points, right?

Reply:
Ok, done!